edn/src/pp.rs

```diff
@@ -0,0 +1,240 @@
+    where T: Into<Option<PrintInstr<'a>>>, U: Into<Option<Vec<PrettyPrintTree<'a>>>> {
+        let instr = instr.into();
+        let children = children.into();
+
+        // Don't be wasteful.
+        assert!(instr.is_some() || children.is_some());
+        assert!(children.is_none() || children.as_ref().unwrap().len() > 0);
+
+        PrettyPrintTree {
+            instr: instr,
+            children: children
+        }
+    }
+
+    /// Recursively consumes this tree and pretty prints it to a string.
+    ///
+    /// # Examples
+    /// ```
+    /// # use std::borrow::Cow;
+    /// # use edn::pp::PrettyPrintTree;
+    /// # use edn::pp::PrintInstr::*;
+    ///
+    /// let tree = PrettyPrintTree::new(PrintString(Cow::from("hello world")), None);
+    /// assert_eq!(tree.pretty_print_into_string().unwrap(), "hello world");
+    /// ```
+    pub fn pretty_print_into_string(self) -> Result<String, io::Error> {
+        self.pretty_print_into_writable(80, Vec::new())
+            .map(|writable| String::from_utf8_lossy(&writable).into_owned())
+    }
+
+    /// Recursively consumes this tree and pretty prints it to memory or any
+    /// given stream.
+    ///
+    /// # Examples
+    /// ```
+    /// # use std::io;
+    /// # use std::borrow::Cow;
+    /// # use edn::pp::PrettyPrintTree;
+    /// # use edn::pp::PrintInstr::*;
+    ///
+    /// let tree = PrettyPrintTree::new(PrintString(Cow::from("hello world")), None);
+    /// assert!(tree.pretty_print_into_writable(80, io::stdout()).is_ok());
+    /// ```
+    pub fn pretty_print_into_writable<W>(self, width: usize, mut out: W) -> Result<W, io::Error> where W: io::Write {
+        let alloc = pretty::BoxAllocator;
+        self.into_doc(&alloc).1.render(width, &mut out).map(|()| out)
+    }
+
+    /// Recursively consumes this tree and creates a pretty.rs document.
+    ///
+    /// # Examples
+    /// ```
+    /// # extern crate pretty;
+    /// # extern crate edn;
+    ///
+    /// # use std::borrow::Cow;
+    /// # use edn::pp::PrettyPrintTree;
+    /// # use edn::pp::PrintInstr::*;
+    ///
+    /// # #[allow(unused_variables)]
+    /// # fn main() {
+    /// let tree = PrettyPrintTree::new(PrintString(Cow::from("hello world")), None);
+    /// let alloc = pretty::BoxAllocator;
+    /// let doc = tree.into_doc(&alloc);
+    /// # }
+    /// ```
+    pub fn into_doc<A>(self, alloc: &'a A) -> pretty::DocBuilder<'a, A>
+    where A: pretty::DocAllocator<'a> {
+        let subdocs = match self.children {
+            Some(v) => v.into_iter().fold(alloc.nil(), |doc, child| doc.append(child.into_doc(alloc))),
+            None => alloc.nil()
+        };
+
+        match self.instr {
+            None => subdocs,
+            Some(PrintInstr::PrintNewline) => alloc.newline().append(subdocs),
+            Some(PrintInstr::PrintString(v)) => alloc.text(v).append(subdocs),
+            Some(PrintInstr::NestChildren(v)) => alloc.newline().append(subdocs).nest(v).append(alloc.newline()),
```

Comment:
@ncalexan Assuming you mean line-by-line streaming to some output, I don't understand how pretty.rs documents can be streamed when rendering needs to happen at the end, so that all the indentation instructions are already in place.

As I've said above, there's a few peculiarities with pretty.rs, where some specific sequences of documents need to be in place for printing to work, especially when handling indented subtrees (at least some of the "unwritten rules" there seemed peculiar to me).